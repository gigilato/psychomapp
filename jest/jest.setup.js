import fetchMock from 'jest-fetch-mock'
import mockSafeAreaContext from 'react-native-safe-area-context/jest/mock'

fetchMock.enableMocks()

jest.useFakeTimers()

require('react-native-reanimated/src/reanimated2/jestUtils').setUpTests()

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

jest.mock('react-native-reanimated', () => {
  return {
    ...require('react-native-reanimated/mock'),
    Layout: {
      easing: jest.fn(),
    },
    useAnimatedKeyboard: () => ({ height: { value: 0 }, state: { value: 0 } }),
  }
})

jest.mock('$atoms/Image/Image.lib', () => ({
  ...jest.requireActual('$atoms/Image/Image.lib'),
  useRatioImage: () => ({
    width: 100,
    height: 100,
  }),
}))

jest.mock('react-native-safe-area-context', () => mockSafeAreaContext)

jest.mock('@react-native-firebase/remote-config', () => ({
  __esModule: true,
  default: () => ({
    setDefaults: jest.fn(),
    getString: jest.fn(),
    getBoolean: jest.fn(),
    getNumber: jest.fn(),
  }),
}))

jest.mock('@react-native-firebase/analytics', () => ({
  __esModule: true,
  default: () => ({
    logEvent: jest.fn(),
    setUserProperty: jest.fn(),
    setUserId: jest.fn(),
  }),
}))

jest.mock('@supabase/supabase-js', () => ({
  createClient: jest.fn(),
}))

jest.mock('react-native-keyboard-controller', () =>
  require('react-native-keyboard-controller/jest')
)

jest.mock('@shopify/react-native-skia', () => {
  const { View } = require('react-native')
  return {
    Skia: {
      Path: {
        Make: () => ({
          moveTo: jest.fn(),
          lineTo: jest.fn(),
          cubicTo: jest.fn(),
          close: jest.fn(),
          addCircle: jest.fn(),
        }),
      },
    },
    vec: jest.fn(),
    Canvas: View,
    Path: View,
    LinearGradient: View,
    Rect: View,
  }
})

jest.mock('$infra/analytics', () => ({
  setConsent: jest.fn(),
  setUser: jest.fn(),
  setUserProperty: jest.fn(),
  setCurrentScreen: jest.fn(),
  clearUser: jest.fn(),
  logEvent: jest.fn(),
  logScreen: jest.fn(),
  useAnalyticsStore: (passedFunction) => {
    const data = {
      currentUserId: undefined,
      currentScreen: '',
      consents: { ga: true, segment: true },
    }
    return passedFunction(data)
  },
}))

jest.mock('$infra/auth', () => ({
  useAuthStore: (passedFunction) => {
    const data = {
      hasRestoredSession: true,
      isAuthenticated: false,
      userId: 'userId',
    }
    return passedFunction(data)
  },
  restoreSessionAsync: jest.fn(),
  signInAsync: jest.fn(),
  signOutAsync: jest.fn(),
  resetPasswordAsync: jest.fn(),
}))

jest.mock('$infra/auth', () => ({
  useAuthStore: (passedFunction) => {
    const data = {
      profile: {
        firstname: 'firstname',
        lastname: 'lastname',
        job: 'job',
        gender: 'male',
        email: 'email@email.com',
      },
      isAuthenticated: false,
      userId: 'userId',
    }
    return passedFunction(data)
  },
  restoreSessionAsync: jest.fn(),
  signInAsync: jest.fn(),
  signOutAsync: jest.fn(),
  resetPasswordAsync: jest.fn(),
}))

function MockSkeleton({ children }) {
  return <>{children}</>
}
MockSkeleton.Group = ({ children }) => <>{children}</>
jest.mock('moti/skeleton', () => ({ Skeleton: MockSkeleton }))

jest.mock('@gorhom/bottom-sheet', () => ({
  __esModule: true,
  ...require('@gorhom/bottom-sheet/mock'),
}))

global.console = {
  ...console,
  // uncomment to ignore a specific log level
  log: jest.fn(),
  debug: jest.fn(),
  info: jest.fn(),
  warn: jest.fn(),
  error: jest.fn(),
}
