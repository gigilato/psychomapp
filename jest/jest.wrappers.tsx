import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { FC } from 'react'

const queryClient = new QueryClient()
export const QueryWrapper: FC<any> = ({ children }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
)
