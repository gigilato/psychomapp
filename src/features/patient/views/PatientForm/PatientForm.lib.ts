import { zodResolver } from '@hookform/resolvers/zod'
import { useRef } from 'react'
import { useForm } from 'react-hook-form'
import { TextInput } from 'react-native'
import { z } from 'zod'

import { useCreatePatient } from '$features/patient/infra/controllers/useCreatePatient'
import { stringSchema } from '$forms/libs/form.schemas'

export const PatientFormSchema = z
  .object({
    firstname: stringSchema,
    lastname: stringSchema,
    birthdate: z.string(),
  })
  .required()

export type PatientFormType = z.infer<typeof PatientFormSchema>

export const usePatientForm = () => {
  const lastnameRef = useRef<TextInput>(null)

  const form = useForm<PatientFormType>({
    resolver: zodResolver(PatientFormSchema),
    defaultValues: {
      firstname: '',
      lastname: '',
      birthdate: '',
    },
    mode: 'onBlur',
  })

  const { mutate, isPending } = useCreatePatient()

  const onPressSubmit = form.handleSubmit((formData) => {
    mutate(formData)
  })

  return { lastnameRef, form, onPressSubmit, isPending }
}
