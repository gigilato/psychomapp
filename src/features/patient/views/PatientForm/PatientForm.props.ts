import { IBoxProps } from '$atoms'

export interface IPatientFormProps extends IBoxProps {
  patientId?: string
  onSubmit: () => void
}
