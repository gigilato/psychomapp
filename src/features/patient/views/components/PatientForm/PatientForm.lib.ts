import { zodResolver } from '@hookform/resolvers/zod'
import { useRef } from 'react'
import { useForm } from 'react-hook-form'
import { TextInput } from 'react-native'
import { z } from 'zod'

import { useCreatePatient } from '$features/patient/infra/controllers/useCreatePatient'
import { usePatient } from '$features/patient/infra/controllers/usePatient'
import { useUpdatePatient } from '$features/patient/infra/controllers/useUpdatePatient'
import { stringSchema } from '$forms/libs/form.schemas'

export const PatientFormSchema = z
  .object({
    firstname: stringSchema,
    lastname: stringSchema,
    birthdate: z.string(),
  })
  .required()

export type PatientFormType = z.infer<typeof PatientFormSchema>

export const usePatientForm = (onSubmit: () => void, patientId?: string) => {
  const lastnameRef = useRef<TextInput>(null)

  const { data: current } = usePatient(patientId)

  const form = useForm<PatientFormType>({
    resolver: zodResolver(PatientFormSchema),
    defaultValues: {
      firstname: current?.firstname ?? '',
      lastname: current?.lastname ?? '',
      birthdate: current?.birthdate ?? '',
    },
    mode: 'onBlur',
  })

  const { mutateAsync: createAsync, isPending: isCreating } = useCreatePatient()
  const { mutateAsync: updateAsync, isPending: isUpdating } = useUpdatePatient()

  const onPressSubmit = form.handleSubmit((formData) => {
    if (patientId) updateAsync({ current: current!, update: formData }).then(() => onSubmit())
    else createAsync(formData).then(() => onSubmit())
  })

  return { lastnameRef, form, onPressSubmit, isPending: patientId ? isUpdating : isCreating }
}
