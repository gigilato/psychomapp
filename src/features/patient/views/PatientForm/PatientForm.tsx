import { Box } from '$atoms'
import { FormDateInput, FormTextInput } from '$forms'
import { i18n } from '$infra/i18n'
import { Button } from '$molecules'

import { PatientFormType, usePatientForm } from './PatientForm.lib'
import { IPatientFormProps } from './PatientForm.props'

export const PatientForm = (props: IPatientFormProps) => {
  const { lastnameRef, form, onPressSubmit, isPending } = usePatientForm()
  return (
    <Box {...props}>
      <FormTextInput<PatientFormType>
        name="firstname"
        label={i18n.t('forms.firstname.label')}
        placeholder={i18n.t('forms.firstname.placeholder')}
        control={form.control}
        type="default"
        onSubmitEditing={() => lastnameRef.current?.focus()}
      />
      <FormTextInput<PatientFormType>
        className="mt-s"
        name="lastname"
        label={i18n.t('forms.lastname.label')}
        placeholder={i18n.t('forms.lastname.placeholder')}
        control={form.control}
        type="default"
        ref={lastnameRef}
      />
      <FormDateInput<PatientFormType> className="mt-s" name="birthdate" control={form.control} />
      <Button
        ID="patientForm/button"
        event="clic_create_patient"
        title={i18n.t('common.validate')}
        onPress={onPressSubmit}
        className="mt-xl"
        isLoading={isPending}
      />
    </Box>
  )
}
