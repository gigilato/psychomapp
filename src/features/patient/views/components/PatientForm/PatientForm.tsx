import { Box } from '$atoms'
import { FormDateInput, FormTextInput } from '$forms'
import { i18n } from '$infra/i18n'
import { Button } from '$molecules'

import { PatientFormType, usePatientForm } from './PatientForm.lib'
import { IPatientFormProps } from './PatientForm.props'

export const PatientForm = ({ patientId, onSubmit, ...props }: IPatientFormProps) => {
  const { lastnameRef, form, onPressSubmit, isPending } = usePatientForm(onSubmit, patientId)
  return (
    <Box {...props}>
      <FormTextInput<PatientFormType>
        name="firstname"
        type="firstname"
        control={form.control}
        onSubmitEditing={() => lastnameRef.current?.focus()}
      />
      <FormTextInput<PatientFormType>
        className="mt-s"
        name="lastname"
        type="lastname"
        control={form.control}
        ref={lastnameRef}
      />
      <FormDateInput<PatientFormType> className="mt-s" name="birthdate" control={form.control} />
      <Button
        ID="patientForm/button"
        event="clic_create_patient"
        title={i18n.t('common.validate')}
        onPress={onPressSubmit}
        className="mt-xl"
        isLoading={isPending}
      />
    </Box>
  )
}
