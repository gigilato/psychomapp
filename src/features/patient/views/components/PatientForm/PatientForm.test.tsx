import { render } from '@testing-library/react-native'

import { QueryWrapper } from '$jest/jest.wrappers'

import { PatientForm } from './PatientForm'

describe('PatientForm', () => {
  it('should render correctly', async () => {
    const Component = render(<PatientForm onSubmit={jest.fn()} />, { wrapper: QueryWrapper })

    expect(Component).toMatchSnapshot()
  })
})
