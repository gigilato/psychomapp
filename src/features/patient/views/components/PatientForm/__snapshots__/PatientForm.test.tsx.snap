// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatientForm should render correctly 1`] = `
<View
  style={
    [
      [
        undefined,
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        [
          undefined,
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "marginBottom": 8,
          },
          {
            "alignItems": "center",
          },
          {
            "justifyContent": "space-between",
          },
          [
            undefined,
            undefined,
          ],
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "SpaceGrotesk-Medium",
            },
            {
              "fontSize": 16,
            },
            {
              "lineHeight": 18,
            },
            {
              "color": "#505050",
            },
          ]
        }
      >
        Prénom
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
          {
            "borderBottomLeftRadius": 6,
            "borderBottomRightRadius": 6,
            "borderTopLeftRadius": 6,
            "borderTopRightRadius": 6,
          },
          {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "overflow": "hidden",
          },
          {
            "borderBottomColor": "#505050",
            "borderLeftColor": "#505050",
            "borderRightColor": "#505050",
            "borderTopColor": "#505050",
          },
          [
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View
        style={
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
            {
              "paddingLeft": 12,
              "paddingRight": 12,
            },
            [
              undefined,
              undefined,
            ],
          ]
        }
      >
        <TextInput
          autoCapitalize="words"
          className="text-[16px] leading-[20px] font-[SpaceGrotesk-Regular] text-[16px] leading-[18px]"
          cursorColor="#FFB300"
          keyboardType="default"
          onBlur={[Function]}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Entrez le prénom"
          selectionColor="#FFB300"
          testID="input/firstname"
          textContentType="givenName"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "marginTop": 12,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "marginBottom": 8,
          },
          {
            "alignItems": "center",
          },
          {
            "justifyContent": "space-between",
          },
          [
            undefined,
            undefined,
          ],
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "SpaceGrotesk-Medium",
            },
            {
              "fontSize": 16,
            },
            {
              "lineHeight": 18,
            },
            {
              "color": "#505050",
            },
          ]
        }
      >
        Nom
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
          {
            "borderBottomLeftRadius": 6,
            "borderBottomRightRadius": 6,
            "borderTopLeftRadius": 6,
            "borderTopRightRadius": 6,
          },
          {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "overflow": "hidden",
          },
          {
            "borderBottomColor": "#505050",
            "borderLeftColor": "#505050",
            "borderRightColor": "#505050",
            "borderTopColor": "#505050",
          },
          [
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View
        style={
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
            {
              "paddingLeft": 12,
              "paddingRight": 12,
            },
            [
              undefined,
              undefined,
            ],
          ]
        }
      >
        <TextInput
          autoCapitalize="words"
          className="text-[16px] leading-[20px] font-[SpaceGrotesk-Regular] text-[16px] leading-[18px]"
          cursorColor="#FFB300"
          keyboardType="default"
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Entrez le nom de famille"
          selectionColor="#FFB300"
          testID="input/lastname"
          textContentType="givenName"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "marginTop": 12,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "marginBottom": 8,
          },
          {
            "alignItems": "center",
          },
          {
            "justifyContent": "space-between",
          },
          [
            undefined,
            undefined,
          ],
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "SpaceGrotesk-Medium",
            },
            {
              "fontSize": 16,
            },
            {
              "lineHeight": 18,
            },
            {
              "color": "#505050",
            },
          ]
        }
      >
        Date de naissance
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
          {
            "borderBottomLeftRadius": 6,
            "borderBottomRightRadius": 6,
            "borderTopLeftRadius": 6,
            "borderTopRightRadius": 6,
          },
          {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
          {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "overflow": "hidden",
          },
          {
            "borderBottomColor": "#505050",
            "borderLeftColor": "#505050",
            "borderRightColor": "#505050",
            "borderTopColor": "#505050",
          },
          [
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View
        style={
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
            {
              "paddingLeft": 12,
              "paddingRight": 12,
            },
            [
              undefined,
              undefined,
            ],
          ]
        }
      >
        <TextInput
          autoCapitalize="none"
          className="text-[16px] leading-[20px] font-[SpaceGrotesk-Regular] text-[16px] leading-[18px]"
          cursorColor="#FFB300"
          editable={false}
          keyboardType="default"
          placeholder="Entrez la date de naissance"
          selectionColor="#FFB300"
          testID="dateInput"
          value=""
        />
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          [
            {
              "opacity": undefined,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ],
        ]
      }
      testID="dateInputPressable"
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    animation="scale"
    collapsable={false}
    event="clic_create_patient"
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "center",
        },
        {
          "justifyContent": "center",
        },
        {
          "borderBottomLeftRadius": 6,
          "borderBottomRightRadius": 6,
          "borderTopLeftRadius": 6,
          "borderTopRightRadius": 6,
        },
        {
          "height": 44,
        },
        {
          "backgroundColor": "#FFB300",
        },
        {
          "marginTop": 32,
        },
        [
          {
            "transform": [
              {
                "scale": undefined,
              },
            ],
          },
          undefined,
        ],
      ]
    }
    testID="patientForm/button"
  >
    <Text
      style={
        [
          {
            "fontFamily": "SpaceGrotesk-Bold",
          },
          {
            "fontSize": 16,
          },
          {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Valider
    </Text>
  </View>
</View>
`;
